src/Planner/CollectTableExpressionData.cpp

```diff
@@ -26,23 +27,37 @@ namespace ErrorCodes
 namespace
 {
 
-class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitor<CollectSourceColumnsVisitor>
+class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitorWithContext<CollectSourceColumnsVisitor>
 {
 public:
     explicit CollectSourceColumnsVisitor(PlannerContextPtr & planner_context_, bool keep_alias_columns_ = true)
-        : planner_context(planner_context_)
+        : InDepthQueryTreeVisitorWithContext(planner_context_->getQueryContext())
+        , planner_context(planner_context_)
         , keep_alias_columns(keep_alias_columns_)
-    {}
+    {
+    }
 
-    void visitImpl(QueryTreeNodePtr & node)
+    void enterImpl(QueryTreeNodePtr & node)
     {
+        if (isIndexHintFunction(node))
+        {
+            is_inside_index_hint_function = true;
+            return;
+        }
+
         auto * column_node = node->as<ColumnNode>();
         if (!column_node)
             return;
 
         if (column_node->getColumnName() == "__grouping_set")
             return;
 
+        /// A special case for the "indexHint" function. We don't need its arguments for execution if column's source table is MergeTree.
+        /// Instead, we prepare an ActionsDAG for its arguments and store it inside a function (see ActionsDAG::buildFilterActionsDAG).
+        /// So this optimization allows not to read arguments of "indexHint" (if not needed in other contexts) but only to use index analysis for them.
+        if (is_inside_index_hint_function && isColumnSourceMergeTree(*column_node))
```

Comment:
Why is it only for MergeTree? What if the source is a subquery? Maybe you don't need to traverse `indexHint` arguments at all?

Reply:
I did it that way initially, but test `02892_orc_filter_pushdown.sql`, which uses `indexHint`, has failed. It happened because storage `File` with `orc` and `parquet` formats doesn't have an index specified in the definition but has an index for all columns, and it can use an index only for columns that are read from the table. However, it doesn't look very useful.

Reply:
This feature is not very important, so let's keep this PR as it is.